paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java

```diff
@@ -85,55 +99,94 @@ public void build() {
         Preconditions.checkNotNull(parserFactory);
 
         StreamExecutionEnvironment env = input.getExecutionEnvironment();
+        if (mode == MySqlDatabaseSyncMode.DYNAMIC) {
+            buildDynamicCdcSink(env);
+        } else {
+            buildStaticCdcSink(env);
+        }
+    }
 
+    private void buildDynamicCdcSink(StreamExecutionEnvironment env) {
+        SingleOutputStreamOperator<Void> parsed =
+                input.forward()
+                        .process(
+                                new CdcDynamicTableParsingProcessFunction<>(
+                                        database, catalogLoader, tables, parserFactory, mode))
+                        .setParallelism(input.getParallelism());
+
+        // for newly-added tables, create a multiplexing operator that handles all their records
+        //     and writes to multiple tables
+        DataStream<CdcMultiplexRecord> newlyAddedTableStream =
+                SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed, CdcDynamicTableParsingProcessFunction.DYNAMIC_OUTPUT_TAG);
+        // handles schema change for newly added tables
+        SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed,
+                        CdcDynamicTableParsingProcessFunction.DYNAMIC_SCHEMA_CHANGE_OUTPUT_TAG)
+                .process(new MultiTableUpdatedDataFieldsProcessFunction(catalogLoader));
+
+        FlinkStreamPartitioner<CdcMultiplexRecord> partitioner =
+                new FlinkStreamPartitioner<>(new CdcMultiplexRecordChannelComputer(catalogLoader));
+        PartitionTransformation<CdcMultiplexRecord> partitioned =
```

Comment:
no one use `partitioned`.

Reply:
used down below:

```java
        if (parallelism != null) {
            partitioned.setParallelism(parallelism);
        }
```

Reply:
I mean what does it do? I see it even has bugs in it, and I'm pretty sure no test can cover it.